### PR TITLE
[BASE-36] Implement AMQP Producer (kombu) prometheus metrics

### PR DESCRIPTION
**Pull request by Xaelias**
## 🧪 Testing Steps / Validation
examples.ini
```
kombu_p.hostname = localhost
kombu_p.exchange_type = topic
kombu_p.exchange_name = test
```

baseplate app
```
with context.span.make_child('kombu', local=True, component_name='stuff') as span:
    try:
        r = span.context.kombu_p.publish("msg", routing_key="123")
    except Exception as e:
        pass
```

/metrics
NOTE: this topic doesn't exist... but as far as I can tell, by default, Kombu doesn't raise this as an error and reports a success
```
# HELP amqp_producer_message_processed_total Multiprocess metric
# TYPE amqp_producer_message_processed_total counter
amqp_producer_message_processed_total{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/"} 1.0
# HELP amqp_producer_message_processing_seconds Multiprocess metric
# TYPE amqp_producer_message_processing_seconds histogram
amqp_producer_message_processing_seconds_sum{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/"} 0.013274467999999984
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.0001"} 0.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.00025"} 0.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.000625"} 0.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.0015625"} 0.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.00390625"} 0.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.009765625"} 0.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.0244140625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.06103515625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.152587890625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.3814697265625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="0.95367431640625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="2.384185791015625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="5.9604644775390625"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="14.901161193847656"} 1.0
amqp_producer_message_processing_seconds_bucket{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/",le="+Inf"} 1.0
amqp_producer_message_processing_seconds_count{amqp_address="localhost:None",amqp_exchange="topic",amqp_queue="test",amqp_success="true",amqp_virtual_host="/"} 1.0
```

## ✅ Checks

- [x] CI tests (if present) are passing
- [x] Adheres to code style for repo
- [x] Contributor License Agreement (CLA) completed if not a Reddit employee

### CODE REVIEW

File: baseplate/clients/kombu.py [hunk 1]
```diff
@@ -11,18 +12,42 @@
...
 
+amqp_producer_labels = [
+    "amqp_address",
+    "amqp_virtual_host",
+    "amqp_exchange",  # exchange_type
+    "amqp_queue",  # exchange_name
+    "amqp_success",
+]
+
+PROCESSING_TIME = Histogram(
```
Comment (KTAtkinson) [inline baseplate/clients/kombu.py]: Should this be `AMQP_PROCESSING_TIME`, as Kombu can be used for more than just reading from AMQP queues?

File: baseplate/clients/kombu.py [hunk 2]
```diff
@@ -232,8 +257,27 @@ def __init__(
         self.exchange = exchange
         self.producers = producers
         self.serializer = serializer
+        self.prom_labels = {
+            "amqp_address": f"{self.connection.hostname}:{self.connection.port}",
```
Comment (KTAtkinson) [inline baseplate/clients/kombu.py:261]: It looks like these labels are being set for all consumers, not just AMQP ones.

**Review by KTAtkinson — CHANGES_REQUESTED** [review]

**Review by KTAtkinson — APPROVED** [review]

File: baseplate/clients/kombu.py [hunk 3]
```diff
@@ -232,8 +257,29 @@ def __init__(
...
+            "amqp_virtual_host": self.connection.virtual_host,
```
Comment (nsheaps) [inline baseplate/clients/kombu.py:262]: should this be added to the spec? Also do you have an example for the spec?

**Review by nsheaps — APPROVED** [review]
Updated [spec](https://github.snooguts.net/reddit/baseplate.spec/pull/51) based on this PR.

We can re-add pool stuff later if it's useful

**Pull request merged**